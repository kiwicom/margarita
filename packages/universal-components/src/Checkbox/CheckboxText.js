// @flow

import * as React from 'react';
import { Text, View, Platform } from 'react-native';

import { StyleSheet } from '../PlatformStyleSheet';
import theme, { parsePxValue } from './styles';

type Props = {|
  +label?: React.Node,
  +checked?: boolean,
  +info?: React.Node,
|};

export default function CheckboxText({ label, checked, info }: Props) {
  return (
    <View style={styles.view}>
      <Text style={[styles.label, checked && styles.labelChecked]}>
        {label}
      </Text>
      <Text style={styles.info}>{info}</Text>
    </View>
  );
}

const styles = StyleSheet.create({
  view: {
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    marginStart: parsePxValue(theme.orbit.spaceXSmall),
  },
  label: {
    fontFamily: Platform.OS === 'web' ? theme.orbit.fontFamily : 'Roboto',
    fontSize: parsePxValue(theme.orbit.fontSizeFormLabel),
    fontWeight: '400',
    color: theme.orbit.colorFormLabel,
    lineHeight: parsePxValue(theme.orbit.heightCheckbox),
  },
  labelChecked: {
    fontWeight: '500',
  },
  info: {
    fontFamily: Platform.OS === 'web' ? theme.orbit.fontFamily : 'Roboto',
    fontSize: parsePxValue(theme.orbit.fontSizeFormFeedback),
    color: theme.orbit.colorInfoCheckBoxRadio,
    lineHeight: 16,
  },
});
