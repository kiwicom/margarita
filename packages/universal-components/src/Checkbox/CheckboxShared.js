// @flow

import * as React from 'react';
import { View, TouchableWithoutFeedback } from 'react-native';
import { defaultTokens } from '@kiwicom/orbit-design-tokens';

import CheckboxText from './CheckboxText';
import CheckboxIcon from './CheckboxIcon';
import { StyleSheet } from '../PlatformStyleSheet';
import { Hoverable } from '../Hoverable';

type Props = {|
  +label?: React.Node,
  +hasError?: boolean,
  +disabled?: boolean,
  +checked?: boolean,
  +info?: React.Node,
  +onPress?: () => void,
  +children?: React.Node,
|};

type State = {|
  focusDisplayed: boolean,
  hovered: boolean,
  pressed: boolean,
|};

class CheckboxShared extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      focusDisplayed: false,
      hovered: false,
      pressed: false,
    };
  }

  handleOnMouseEnter = () => {
    this.setState({ hovered: true });
  };

  handleOnMouseLeave = () => {
    this.setState({ hovered: false });
  };

  handleOnPressIn = () => {
    this.setState({ pressed: true });
  };

  handleOnPressOut = () => {
    this.setState({ pressed: false, focusDisplayed: false });
  };

  handleOnFocus = () => {
    this.setState({ focusDisplayed: true });
  };

  handleOnBlur = () => {
    this.setState({ focusDisplayed: false });
  };

  render() {
    const {
      label,
      hasError,
      disabled,
      checked,
      info,
      onPress,
      children,
    } = this.props;
    const { focusDisplayed, hovered, pressed } = this.state;

    return (
      <Hoverable
        onMouseEnter={this.handleOnMouseEnter}
        onMouseLeave={this.handleOnMouseLeave}
      >
        <TouchableWithoutFeedback
          accessibilityRole="button"
          onPress={onPress}
          onPressIn={this.handleOnPressIn}
          onPressOut={this.handleOnPressOut}
          onFocus={this.handleOnFocus}
          onBlur={this.handleOnBlur}
          disabled={disabled}
        >
          <View style={[styles.view, disabled && styles.viewDisabled]}>
            <CheckboxIcon
              checked={checked}
              hasError={hasError}
              disabled={disabled}
              focused={focusDisplayed}
              hovered={hovered}
              pressed={pressed}
            />
            {children || (
              <CheckboxText label={label} checked={checked} info={info} />
            )}
          </View>
        </TouchableWithoutFeedback>
      </Hoverable>
    );
  }
}

const styles = StyleSheet.create({
  view: {
    display: 'flex',
    flexDirection: 'row',
    web: {
      outline: '0',
    },
  },
  viewDisabled: {
    opacity: parseFloat(defaultTokens.opacityCheckboxDisabled),
  },
});

export default CheckboxShared;
