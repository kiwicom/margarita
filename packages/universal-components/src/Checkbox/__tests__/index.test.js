// @flow

import * as React from 'react';
import { Platform, View } from 'react-native';
import { render, fireEvent } from 'react-native-testing-library';
import snapshotDiff from 'snapshot-diff';

import CheckboxWeb from '../Checkbox.web';
import CheckboxNative from '../Checkbox.native';
import CheckboxShared from '../CheckboxShared';
import CheckboxText from '../CheckboxText';

const originalPlatform = Platform.OS;
const label = 'Label';
const value = 'value';
const hasError = false;
const disabled = false;
const checked = true;
const name = 'name';
const info = 'info';
const onChange = jest.fn();

afterEach(() => {
  Platform.OS = originalPlatform;
});

describe('CheckBox - web', () => {
  Platform.OS = 'web';
  const { getByType, getByText, getAllByProps } = render(
    <CheckboxWeb
      label={label}
      value={value}
      hasError={hasError}
      disabled={disabled}
      checked={checked}
      name={name}
      info={info}
      onChange={onChange}
    />,
  );

  it('should contain a label', () => {
    expect(getByText(label)).toBeDefined();
  });

  it('should contain a info', () => {
    expect(getByText(info)).toBeDefined();
  });

  it('should have passed props', () => {
    expect(
      getAllByProps({
        label,
        value,
        hasError,
        disabled,
        checked,
        name,
        info,
        onChange,
      }),
    ).toBeDefined();
  });

  it('should execute onChange method', () => {
    fireEvent(getByType(CheckboxWeb), 'change');
  });

  it('should match snapshot diff', () => {
    const base = render(<CheckboxWeb />);
    const extend = render(
      <CheckboxWeb
        label={label}
        value={value}
        hasError={hasError}
        disabled={disabled}
        checked={checked}
        name={name}
        info={info}
        onChange={onChange}
      />,
    );

    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });
});

describe('CheckBox - native', () => {
  Platform.OS = 'ios';
  const { getByType, getByText, getAllByProps } = render(
    <CheckboxNative
      label={label}
      hasError={hasError}
      disabled={disabled}
      checked={checked}
      info={info}
      onChange={onChange}
    />,
  );

  it('should contain a label', () => {
    expect(getByText(label)).toBeDefined();
  });

  it('should contain a info', () => {
    expect(getByText(info)).toBeDefined();
  });

  it('should have passed props', () => {
    expect(
      getAllByProps({
        label,
        hasError,
        disabled,
        checked,
        info,
        onChange,
      }),
    ).toBeDefined();
  });

  it('should execute onChange method', () => {
    fireEvent(getByType(CheckboxNative), 'change');
  });

  it('should match snapshot diff', () => {
    const base = render(<CheckboxNative />);
    const extend = render(
      <CheckboxNative
        label={label}
        hasError={hasError}
        disabled={disabled}
        checked={checked}
        info={info}
        onChange={onChange}
      />,
    );

    expect(snapshotDiff(base, extend)).toMatchSnapshot();
  });
});

describe('CheckboxShared', () => {
  it('should render CheckboxText with label', () => {
    const wrapper = render(<CheckboxShared label="label" />);
    expect(wrapper.getByType(CheckboxText)).toBeDefined();
  });

  it('should render children instead of label', () => {
    const wrapper = render(
      <CheckboxShared label="label">
        <View testID="test-view" />
      </CheckboxShared>,
    );
    expect(wrapper.getByTestId('test-view')).toBeDefined();
    expect(() => {
      wrapper.getByType(CheckboxText);
    }).toThrow();
  });
});
