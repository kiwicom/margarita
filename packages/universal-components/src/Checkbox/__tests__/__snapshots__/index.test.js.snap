// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBox - native should match snapshot diff 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -11,11 +11,11 @@
      Array [
        Object {
          \\"display\\": \\"flex\\",
          \\"flexDirection\\": \\"row\\",
        },
-       undefined,
+       false,
      ]
    }
  >
    <View
      style={
@@ -29,17 +29,39 @@
            \\"borderWidth\\": 1,
            \\"height\\": 20,
            \\"justifyContent\\": \\"center\\",
            \\"width\\": 20,
          },
-         undefined,
          false,
          false,
          false,
+         false,
        ]
      }
-   />
+   >
+     <Text
+       style={
+         Array [
+           Object {
+             \\"fontFamily\\": \\"orbit-icons\\",
+           },
+           Object {
+             \\"fontSize\\": 16,
+             \\"height\\": 16,
+             \\"lineHeight\\": 16,
+             \\"width\\": 16,
+           },
+           Object {
+             \\"color\\": \\"#00a991\\",
+           },
+           undefined,
+         ]
+       }
+     >
+       V
+     </Text>
+   </View>
    <View
      style={
        Object {
          \\"display\\": \\"flex\\",
          \\"flex\\": 1,
@@ -56,21 +78,27 @@
              \\"fontFamily\\": \\"Roboto\\",
              \\"fontSize\\": 14,
              \\"fontWeight\\": \\"400\\",
              \\"lineHeight\\": 20,
            },
-           undefined,
+           Object {
+             \\"fontWeight\\": \\"500\\",
+           },
          ]
        }
-     />
+     >
+       Label
+     </Text>
      <Text
        style={
          Object {
            \\"color\\": \\"#7f91a8\\",
            \\"fontFamily\\": \\"Roboto\\",
            \\"fontSize\\": 12,
            \\"lineHeight\\": 16,
          }
        }
-     />
+     >
+       info
+     </Text>
    </View>
  </View>"
`;

exports[`CheckBox - web should match snapshot diff 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -6,17 +6,21 @@
        \\"flexDirection\\": \\"row\\",
      }
    }
  >
    <input
+     checked={true}
+     disabled={false}
+     name=\\"name\\"
      style={
        Object {
          \\"display\\": \\"none\\",
          \\"visibility\\": \\"hidden\\",
        }
      }
      type=\\"checkbox\\"
+     value=\\"value\\"
    />
    <View
      accessibilityRole=\\"button\\"
      accessible={true}
      onResponderGrant={[Function bound touchableHandleResponderGrant]}
@@ -29,11 +33,11 @@
        Array [
          Object {
            \\"display\\": \\"flex\\",
            \\"flexDirection\\": \\"row\\",
          },
-         undefined,
+         false,
        ]
      }
    >
      <View
        style={
@@ -47,17 +51,39 @@
              \\"borderWidth\\": 1,
              \\"height\\": 20,
              \\"justifyContent\\": \\"center\\",
              \\"width\\": 20,
            },
-           undefined,
+           false,
            false,
            false,
            false,
          ]
        }
-     />
+     >
+       <Text
+         style={
+           Array [
+             Object {
+               \\"fontFamily\\": \\"orbit-icons\\",
+             },
+             Object {
+               \\"fontSize\\": 16,
+               \\"height\\": 16,
+               \\"lineHeight\\": 16,
+               \\"width\\": 16,
+             },
+             Object {
+               \\"color\\": \\"#00a991\\",
+             },
+             undefined,
+           ]
+         }
+       >
+         V
+       </Text>
+     </View>
      <View
        style={
          Object {
            \\"display\\": \\"flex\\",
            \\"flex\\": 1,
@@ -74,22 +100,28 @@
                \\"fontFamily\\": \\"Roboto\\",
                \\"fontSize\\": 14,
                \\"fontWeight\\": \\"400\\",
                \\"lineHeight\\": 20,
              },
-             undefined,
+             Object {
+               \\"fontWeight\\": \\"500\\",
+             },
            ]
          }
-       />
+       >
+         Label
+       </Text>
        <Text
          style={
            Object {
              \\"color\\": \\"#7f91a8\\",
              \\"fontFamily\\": \\"Roboto\\",
              \\"fontSize\\": 12,
              \\"lineHeight\\": 16,
            }
          }
-       />
+       >
+         info
+       </Text>
      </View>
    </View>
  </label>"
`;
